### PR TITLE
fix(tianmu) Query results more data (#282)

### LINKED ISSUE #282: bug:  incorrect result set.Query results more data, see the bug text
### Describe the problem

```
CREATE TABLE t1 (sku int PRIMARY KEY, pr int)engine=stonedb;
CREATE TABLE t2 (sku int PRIMARY KEY, sppr int, name varchar(255))engine=stonedb;

INSERT INTO t1 VALUES
  (10, 10), (20, 10), (30, 20), (40, 30), (50, 10), (60, 10);

INSERT INTO t2 VALUES 
  (10, 10, 'aaa'), (20, 10, 'bbb'), (30, 10, 'ccc'), (40, 20, 'ddd'),
  (50, 10, 'eee'), (60, 20, 'fff'), (70, 20, 'ggg'), (80, 30, 'hhh');

SELECT t2.sku, t2.sppr, t2.name, t1.sku, t1.pr
  FROM t2, t1 WHERE t2.sku=20 AND (t2.sku=t1.sku OR t2.sppr=t1.sku);

##The error results are as follows
+-----+------+------+-----+------+
| sku | sppr | name | sku | pr   |
+-----+------+------+-----+------+
|  20 |   10 | bbb  |  10 |   10 |
|  20 |   10 | bbb  |  20 |   10 |
|  20 |   10 | bbb  |  30 |   20 |
|  20 |   10 | bbb  |  40 |   30 |
|  20 |   10 | bbb  |  50 |   10 |
|  20 |   10 | bbb  |  60 |   10 |
+-----+------+------+-----+------+
6 rows in set (0.02 sec)
```

### Expected behavior

```
##The following is：innodb test result

+-----+------+------+-----+------+
| sku | sppr | name | sku | pr   |
+-----+------+------+-----+------+
|  20 |   10 | bbb  |  10 |   10 |
|  20 |   10 | bbb  |  20 |   10 |
+-----+------+------+-----+------+
2 rows in set (0.00 sec)
```

### How To Reproduce

```
CREATE TABLE t1 (sku int PRIMARY KEY, pr int)engine=stonedb;
CREATE TABLE t2 (sku int PRIMARY KEY, sppr int, name varchar(255))engine=stonedb;

INSERT INTO t1 VALUES
  (10, 10), (20, 10), (30, 20), (40, 30), (50, 10), (60, 10);

INSERT INTO t2 VALUES 
  (10, 10, 'aaa'), (20, 10, 'bbb'), (30, 10, 'ccc'), (40, 20, 'ddd'),
  (50, 10, 'eee'), (60, 20, 'fff'), (70, 20, 'ggg'), (80, 30, 'hhh');

SELECT t2.sku, t2.sppr, t2.name, t1.sku, t1.pr
  FROM t2, t1 WHERE t2.sku=20 AND (t2.sku=t1.sku OR t2.sppr=t1.sku);
```

### Environment

1. StoneDB for mysql5.7 (release)
2. Ubuntu 20.04.4

### Are you interested in submitting a PR to solve the problem?

- [x] Yes, I will!

### PR DESCRIPTION
**Pull request by konghaiya**
Cause:
Execution plan error when AND/OR sql logical operator in condition occurs at the same time.
Resolvent:
Optimize multiple equal execution plan logic.

## Summary about this PR

Issue Number: close #282

## Tests Check List

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [x] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: mysql-test/suite/tianmu/r/issue282.result [hunk 1]
```diff
@@ -0,0 +1,83 @@
+use test;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/r/issue282.result:1]: using `include have_tianmu.inc` ?

Comment (RingsC) [reply mysql-test/suite/tianmu/r/issue282.result:1]: disable query log or not?

Comment (RingsC) [reply mysql-test/suite/tianmu/r/issue282.result:1]: more charset testing? unicode or utf-8, etc.

**Review by isredstar — APPROVED** [review]
LGTM

**Review by RingsC — CHANGES_REQUESTED** [review]

File: mysql-test/suite/tianmu/r/issue282.result (hunk 1 above)
Comment (konghaiya) [reply mysql-test/suite/tianmu/r/issue282.result:1]: Yes, we have used (have_tianmu. inc) to check before testing.

File: storage/tianmu/core/query.cpp [hunk 2]
```diff
@@ -1122,33 +1122,27 @@ CondID Query::ConditionNumberFromMultipleEquality(Item_equal *conds, const TabID
   ifield = li++;
   if (!Item2CQTerm(ifield, first_term, tmp_table, filter_type)) return CondID(-1);
   CondID filter;
```
Comment (RingsC) [inline storage/tianmu/core/query.cpp:1127]: Refine these code.
```
  Item_field *ifield;
  Item *const_item = conds->get_const();
  if (const_item) {
    if (!Item2CQTerm(const_item, zero_term, tmp_table, filter_type)) return CondID(-1);
  } else {
    ifield = li++;
    if (!Item2CQTerm(ifield, zero_term, tmp_table, filter_type)) return CondID(-1);
  }
wired codes. 

can change to :
  Item_field *ifield;
  Item *const_item = conds->get_const();
  if (const_item) {
    if (!Item2CQTerm(const_item, zero_term, tmp_table, filter_type)) return CondID(-1);

 ifield = li++;
 if (!Item2CQTerm(ifield, zero_term, tmp_table, filter_type)) return CondID(-1);
```
and why use a particular line code to deal with `const_item` , const_item is also an item of equlality expression.


`
/*
  The class Item_equal is used to represent conjunctions of equality
  predicates of the form field1 = field2, and field=const in where
  conditions and on expressions.

  All equality predicates of the form field1=field2 contained in a
  conjunction are substituted for a sequence of items of this class.
  An item of this class Item_equal(f1,f2,...fk) represents a
  multiple equality f1=f2=...=fk.

  If a conjunction contains predicates f1=f2 and f2=f3, a new item of
  this class is created Item_equal(f1,f2,f3) representing the multiple
  equality f1=f2=f3 that substitutes the above equality predicates in
  the conjunction.
  A conjunction of the predicates f2=f1 and f3=f1 and f3=f2 will be
  substituted for the item representing the same multiple equality
  f1=f2=f3.
  An item Item_equal(f1,f2) can appear instead of a conjunction of 
  f2=f1 and f1=f2, or instead of just the predicate f1=f2.

  An item of the class Item_equal inherits equalities from outer 
  conjunctive levels.

  Suppose we have a where condition of the following form:
  WHERE f1=f2 AND f3=f4 AND f3=f5 AND ... AND (...OR (f1=f3 AND ...)).
  In this case:
    f1=f2 will be substituted for Item_equal(f1,f2);
    f3=f4 and f3=f5  will be substituted for Item_equal(f3,f4,f5);
    f1=f3 will be substituted for Item_equal(f1,f2,f3,f4,f5);

  An object of the class Item_equal can contain an optional constant
  item c. Then it represents a multiple equality of the form 
  c=f1=...=fk.

  Objects of the class Item_equal are used for the following:

  1. An object Item_equal(t1.f1,...,tk.fk) allows us to consider any
  pair of tables ti and tj as joined by an equi-condition.
  Thus it provide us with additional access paths from table to table.

  2. An object Item_equal(t1.f1,...,tk.fk) is applied to deduce new
  SARGable predicates:
    f1=...=fk AND P(fi) => f1=...=fk AND P(fi) AND P(fj).
  It also can give us additional index scans and can allow us to
  improve selectivity estimates.

  3. An object Item_equal(t1.f1,...,tk.fk) is used to optimize the 
  selected execution plan for the query: if table ti is accessed 
  before the table tj then in any predicate P in the where condition
  the occurrence of tj.fj is substituted for ti.fi. This can allow
  an evaluation of the predicate at an earlier step.

  When feature 1 is supported they say that join transitive closure 
  is employed.
  When feature 2 is supported they say that search argument transitive
  closure is employed.
  Both features are usually supported by preprocessing original query and
  adding additional predicates.
  We do not just add predicates, we rather dynamically replace some
  predicates that can not be used to access tables in the investigated
  plan for those, obtained by substitution of some fields for equal fields,
  that can be used.     

  Prepared Statements/Stored Procedures note: instances of class
  Item_equal are created only at the time a PS/SP is executed and
  are deleted in the end of execution. All changes made to these
  objects need not be registered in the list of changes of the parse
  tree and do not harm PS/SP re-execution.

  Item equal objects are employed only at the optimize phase. Usually they are
  not supposed to be evaluated.  Yet in some cases we call the method val_int()
  for them. We have to take care of restricting the predicate such an
  object represents f1=f2= ...=fn to the projection of known fields fi1=...=fik.
*/
`

Comment (konghaiya) [reply storage/tianmu/core/query.cpp:1127]: ACK.
I need to analyze it again.

File: storage/tianmu/core/query.cpp [hunk 3]
```diff
@@ -1122,33 +1122,27 @@ CondID Query::ConditionNumberFromMultipleEquality(Item_equal *conds, const TabID
...
-  if (!and_me_filter)
```
Comment (RingsC) [inline storage/tianmu/core/query.cpp:1125]: How about if the  expression with `and, or `, such as `( a and (b or c))` ?

File: storage/tianmu/core/query.cpp [hunk 4]
```diff
@@ -1122,33 +1122,27 @@ CondID Query::ConditionNumberFromMultipleEquality(Item_equal *conds, const TabID
...
-  else {
-    if (is_or_subtree)
-      cq->Or(*and_me_filter, tmp_table, first_term, common::Operator::O_EQ, zero_term);
-    else
-      cq->And(*and_me_filter, tmp_table, first_term, common::Operator::O_EQ, zero_term);
-  }
+
   while ((ifield = li++) != nullptr) {
     if (!Item2CQTerm(ifield, next_term, tmp_table, filter_type)) return CondID(-1);
-    if (!and_me_filter) {
```
Comment (RingsC) [inline storage/tianmu/core/query.cpp:1136]: as said in 
```
        while ((item = li++) != nullptr) {
          CondID res = ConditionNumber(item, tmp_table, filter_type, and_cond.get(),
                                       // if there is no and_cond the filter has to be created and info
                                       // about tree/no tree like descriptor has to be passed recursively
                                       // down once filter is created we pass 'false' to indicate AND
                                       and_cond.get() ? false : is_or_subtree);
```


To processing OR claus.
```
        li.rewind();
        // generate non-transformed conditions
        while ((item = li++) != nullptr) {
          if (is_transformed[item] == true) continue;
          create_or_subtree = true;
          CondID res = ConditionNumber(item, tmp_table, filter_type, or_cond.get(), true /*CondType::OR_SUBTREE*/);
          if (res.IsInvalid()) return res;
          if (!or_cond) or_cond.reset(new CondID(res.n));
        }
```

`is_or_subtree` is maybe true, here 
```
  while ((ifield = li++) != nullptr) {
    if (!Item2CQTerm(ifield, next_term, tmp_table, filter_type)) return CondID(-1);
    cq->And(filter, tmp_table, next_term, common::Operator::O_EQ, zero_term);
```
we just only generate an `And` compiled query. why?

File: storage/tianmu/core/query.cpp [hunk 5]
```diff
@@ -1122,33 +1122,27 @@ CondID Query::ConditionNumberFromMultipleEquality(Item_equal *conds, const TabID
...
+
+  if (and_me_filter) {
+    if (is_or_subtree) {
+      cq->Or(*and_me_filter, tmp_table, filter);
     } else {
-      if (is_or_subtree)
-        cq->Or(*and_me_filter, tmp_table, next_term, common::Operator::O_EQ, zero_term);
-      else
-        cq->And(*and_me_filter, tmp_table, next_term, common::Operator::O_EQ, zero_term);
+      cq->And(*and_me_filter, tmp_table, filter);
```
Comment (RingsC) [inline storage/tianmu/core/query.cpp:1139]: See comments

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/query.cpp (hunk 4 above)
Comment (konghaiya) [reply storage/tianmu/core/query.cpp:1136]: This function deals with equivalence optimization. Equivalence or constant transfer optimization will convert multiple conditions into one condition, so use and.

**Review by RingsC — APPROVED** [review]
LGTM
